mpc85xx: add support for Extreme Networks WS-AP3825i

Hurricos opened the pull request:
Hardware:

- SoC:     Freescale P1020
  - CPU:     2x e500v2 @ 800MHz
- Flash:   64MiB NOR (1x Intel JS28F512)
- Memory:  256MiB (2x ProMOS DDR3 V73CAG01168RBJ-I9H 1Gb)
- WiFi1:   2.4+5GHz abgn 3x3 (Atheros AR9590)
- Wifi2:   5GHz an+ac 3x3 (Qualcomm Atheros QCA9890)
- ETH:     2x PoE Gigabit Ethernet (2x Atheros AR8035)
- Power:   12V (center-positive barrel) or 48V PoE (active or passive)
- Serial:  Cisco-compatible RJ45 next to 12V power socket (115200 baud)
- LED Driver: TI LV164A
  - LEDs: (not functioning)
    - 2x Power (Green + Orange)
    - 4x ETH (ETH1 + ETH2) x (Green + Orange)
    - 2x WiFi (WiFi2 + WiFi1)

Installation:

1. Grab the OpenWrt initramfs <openwrt-initramfs-bin>, e.g.
   openwrt-mpc85xx-p1020-extreme-networks_ws-ap3825i-initramfs-kernel.bin.
   Place it in the root directory of a DHCP+TFTP server, e.g. OpenWrt
   `dnsmasq` with configuration `dhcp.server.enable_tftp='1'`.

2. Connect to the serial port and boot the AP with options
   e.g. 115200,N,8. Stop autoboot in U-Boot by pressing Enter after
   'Scanning JFFS2 FS:' begins, then waiting for the prompt to be
   interrupted. Credentials are identical to the one in the APs
   interface. By default it is admin / new2day: if these do not work,
   follow the OEM's reset procedure using the reset button.

3. Set the bootcmd so the AP can boot OpenWrt by executing:

```uboot
setenv boot_openwrt "cp.b 0xEC000000 0x2000000 0x2000000; interrupts off; bootm start 0x2000000; bootm loados; fdt resize; fdt boardsetup; fdt resize; fdt boardsetup; fdt chosen; fdt resize; fdt chosen; bootm prep; bootm go;"
setenv bootcmd "run boot_openwrt"
saveenv
```

   If you plan on going back to the vendor firmware - the bootcmd for it
   is stored in the boot_flash variable.

4. Load the initramfs image to RAM and boot by executing (you can do this in two lines of copy-pasting):

```uboot
setenv ipaddr <ipv4 client address>; setenv serverip <tftp server address>; tftpboot 0x2000000 <openwrt-initramfs-bin>;
interrupts off; bootm start 0x2000000; bootm loados; fdt resize; fdt boardsetup; fdt resize; fdt boardsetup; fdt chosen; fdt resize; fdt chosen; bootm prep; bootm go; 
```

5. Make a backup of the "firmware" partition if you ever wish to go back
   to the vendor firmware.

6. Upload the OpenWrt sysupgrade image via SCP to the devices /tmp
   folder.

7. Flash OpenWrt using sysupgrade.

```ash
sysupgrade /tmp/<openwrt-sysupgrade-bin>
```

Notes:

- We must step through the `bootm` process manually to avoid fdt
  relocation. To explain: the stock U-boot (and stock Linux) are configured
  with a very large CONFIG_SYS_BOOTMAPSZ (and the device's stock Linux
  kernel is configured to be able to handle it). The U-boot version
  predates the check for the `fdt_high` variable, meaning that upon fdt
  relocation, the fdt can (and will) be moved to a very high address; the
  default appears to be 0x9ffa000. This address is so high that when the
  Linux kernel starts reading the fdt at the beginning of the boot process,
  it encounters a memory access exception and panics[5]. While it is
  possible to reduce the highest address the fdt will be relocated to by
  setting `bootm_size`, this also has the side effect of limiting the
  amount of RAM the kernel can use[3].

- The flattened device tree must be resized multiple times to deal with
  U-boot's pickiness about making space for the fdt without relocating
  it. Although `help fdt` advertises that `fdt resize` pads to a 4K-aligned
  address, it cannot predict the future size of the fdt after later `fdt`
  operations. To futureproof against later changes to the dtb and the real
  alignment of the fdt within the fit image, perform each `fdt` operation
  twice, once before and once after a separate `fdt resize`; thankfully,
  these `fdt` operations are idempotent on their own.

- The primary ethernet MAC address is stored (and set) in U-boot; they are
  shimmed into the device tree by 'fdt boardsetup' through the
  'local-mac-address' property of the respective ethernet node, so OpenWrt
  does not need to set this at runtime. Note that U-boot indexes the
  ethernet nodes by alias, which is why the device tree explicitly aliases
  ethernet1 to enet2.

- LEDs do not function under OpenWrt. Each of 8 LEDs is connected to an
  output of a TI LV164A shift register, which is wired to GPIO lines and
  operates through bit-banged SPI. Unfortunately, I am unable to get the
  spi-gpio driver to recognize the `led_spi` device tree node at all, as
  confirmed by patching in printk messages demonstrating
  spi-gpio.c::spi_gpio_probe never runs. It is possible to manually
  articulate the shift register by exporting the GPIO lines and stepping
  their values through the sysfs.

- Though they do not function under OpenWrt, I have left the pinout details
  of the LEDs and shift register in the device tree to represent real
  hardware.

- An archive of the u-boot and Linux source for the AP3825i (which is one
  device of a range of devices code-named 'CHANTRY') be found here:
  magnet:?xt=urn:btih:f5306a5dfd06d42319e4554565429f84dde96bbc

- The device has an identical case to both the Enterasys WS-AP3715i and
  Adtran BSAP-2030[2] (and potentially other Adtran BSAPs). Given that
  there is no FCC ID for the board itself (only its WLAN modules), it's
  likely these are generic boards, and even that the WS-AP3715i is
  identical, with only a change in WLAN card. I have ordered one to confirm
  this.

- For additional information: the process of porting the board is
  documented in an OpenWrt forum thread[4].

[2]: https://forum.openwrt.org/t/support-for-adtran-bluesocket-bsap-2030/48538
[3]: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168/29
[4]: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168
[5]: https://paste.c-net.org/LakhiToner

Tested-by: Martin Kennedy <hurricos@gmail.com>
Signed-off-by: Martin Kennedy <hurricos@gmail.com>

Hurricos committed:
mpc85xx: add support for Extreme Networks WS-AP3825i

Hardware:

- SoC:     Freescale P1020
  - CPU:     2x e500v2 @ 800MHz
- Flash:   64MiB NOR (1x Intel JS28F512)
- Memory:  256MiB (2x ProMOS DDR3 V73CAG01168RBJ-I9H 1Gb)
- WiFi1:   2.4+5GHz abgn 3x3 (Atheros AR9590)
- Wifi2:   5GHz an+ac 3x3 (Qualcomm Atheros QCA9890)
- ETH:     2x PoE Gigabit Ethernet (2x Atheros AR8035)
- Power:   12V (center-positive barrel) or 48V PoE (active or passive)
- Serial:  Cisco-compatible RJ45 next to 12V power socket (115200 baud)
- LED Driver: TI LV164A
  - LEDs: (not functioning)
    - 2x Power (Green + Orange)
    - 4x ETH (ETH1 + ETH2) x (Green + Orange)
    - 2x WiFi (WiFi2 + WiFi1)

Installation:

1. Grab the OpenWrt initramfs <openwrt-initramfs-bin>, e.g.
   openwrt-mpc85xx-p1020-extreme-networks_ws-ap3825i-initramfs-kernel.bin.
   Place it in the root directory of a DHCP+TFTP server, e.g. OpenWrt
   `dnsmasq` with configuration `dhcp.server.enable_tftp='1'`.

2. Connect to the serial port and boot the AP with options
   e.g. 115200,N,8. Stop autoboot in U-Boot by pressing Enter after
   'Scanning JFFS2 FS:' begins, then waiting for the prompt to be
   interrupted. Credentials are identical to the one in the APs
   interface. By default it is admin / new2day: if these do not work,
   follow the OEM's reset procedure using the reset button.

3. Set the bootcmd so the AP can boot OpenWrt by executing:

```uboot
setenv boot_openwrt "cp.b 0xEC000000 0x2000000 0x2000000; interrupts off; bootm start 0x2000000; bootm loados; fdt resize; fdt boardsetup; fdt chosen; bootm prep; bootm go;"
setenv bootcmd "run boot_openwrt"
saveenv
```

   If you plan on going back to the vendor firmware - the bootcmd for it
   is stored in the boot_flash variable.

4. Load the initramfs image to RAM and boot by executing

```uboot
setenv ipaddr <ipv4 client address>;
setenv serverip <tftp server address>;
tftpboot 0x2000000 <openwrt-initramfs-bin>;
interrupts off;
bootm start 0x2000000;
bootm loados;
fdt resize;
fdt boardsetup;
fdt chosen;
bootm prep;
bootm go;
```

5. Make a backup of the "firmware" partition if you ever wish to go back
   to the vendor firmware.

6. Upload the OpenWrt sysupgrade image via SCP to the devices /tmp
   folder.

7. Flash OpenWrt using sysupgrade.

```ash
sysupgrade /tmp/<openwrt-sysupgrade-bin>
```

Notes:

- We must step through the `bootm` process manually to avoid fdt
  relocation. To explain: the stock U-boot (and stock Linux) are configured
  with a very large CONFIG_SYS_BOOTMAPSZ (and the device's stock Linux
  kernel is configured to be able to handle it). The U-boot version
  predates the check for the `fdt_high` variable, meaning that upon fdt
  relocation, the fdt can (and will) be moved to a very high address; the
  default appears to be 0x9ffa000. This address is so high that when the
  Linux kernel starts reading the fdt at the beginning of the boot process,
  it encounters a memory access exception and panics[5]. While it is
  possible to reduce the highest address the fdt will be relocated to by
  setting `bootm_size`, this also has the side effect of limiting the
  amount of RAM the kernel can use[3].

- Because it is not relocated, the flattened device tree needs to be
  padded in the build process to guarantee that `fdt resize` has
  enough space.

- The primary ethernet MAC address is stored (and set) in U-boot; they are
  shimmed into the device tree by 'fdt boardsetup' through the
  'local-mac-address' property of the respective ethernet node, so OpenWrt
  does not need to set this at runtime. Note that U-boot indexes the
  ethernet nodes by alias, which is why the device tree explicitly aliases
  ethernet1 to enet2.

- LEDs do not function under OpenWrt. Each of 8 LEDs is connected to an
  output of a TI LV164A shift register, which is wired to GPIO lines and
  operates through bit-banged SPI. Unfortunately, I am unable to get the
  spi-gpio driver to recognize the `led_spi` device tree node at all, as
  confirmed by patching in printk messages demonstrating
  spi-gpio.c::spi_gpio_probe never runs. It is possible to manually
  articulate the shift register by exporting the GPIO lines and stepping
  their values through the sysfs.

- Though they do not function under OpenWrt, I have left the pinout details
  of the LEDs and shift register in the device tree to represent real
  hardware.

- An archive of the u-boot and Linux source for the AP3825i (which is one
  device of a range of devices code-named 'CHANTRY') be found here[1].

- The device has an identical case to both the Enterasys WS-AP3725i and
  Adtran BSAP-2030[2] (and potentially other Adtran BSAPs). Given that
  there is no FCC ID for the board itself (only its WLAN modules), it's
  likely these are generic boards, and even that the WS-AP3725i is
  identical, with only a change in WLAN card. I have ordered one to confirm
  this.

- For additional information: the process of porting the board is
  documented in an OpenWrt forum thread[4].

[1]: magnet:?xt=urn:btih:f5306a5dfd06d42319e4554565429f84dde96bbc
[2]: https://forum.openwrt.org/t/support-for-adtran-bluesocket-bsap-2030/48538
[3]: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168/29
[4]: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168
[5]: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168/26

Tested-by: Martin Kennedy <hurricos@gmail.com>
Signed-off-by: Martin Kennedy <hurricos@gmail.com>

Hurricos commented:
Hey @Gadorach and @voronin10: Would you be willing to test your boards at this exact commit (for convenience, I have kicked off a build to be published [here](https://downloads.laboratoryb.org/releases/snapshot/targets/mpc85xx/p1020/) which will finish around 22:10 UTC)? If so, could you then comment "Tested by: Your Name <your.email.address@example.com>" so I may add it to my commit?

Thank you both for testing my port of this device :D

voronin10 commented:
> Hey @Gadorach and @voronin10: Would you be willing to test your boards at this exact commit (for convenience, I have kicked off a build to be published [here](https://downloads.laboratoryb.org/releases/snapshot/targets/mpc85xx/p1020/) which will finish around 22:10 UTC)? If so, could you then comment "Tested by: Your Name [your.email.address@example.com](mailto:your.email.address@example.com)" so I may add it to my commit?
> 
> Thank you both for testing my port of this device :D

Hi, I good test your build on my 2 devices,  it works stable many time
Lan1 Lan2 ports are good works same  wireless
I use this device just a dumb AP

voronin10 commented:
I have a problem
i use this
```

------------clone
git clone git://github.com/Hurricos/openwrt.git  -b add_ap3825i 
cd ~/openwrt
./scripts/feeds update -a && ./scripts/feeds install -a 
 make prereq
------------after 
Target system -> mpc85xx
Subtarget -> P1020
Target-> 3825i
Luci -> Collections -> luci
----exit
make -j5 V=s

```
but if I make Openwrt with luci, terminal return error
```

pkg_hash_check_unresolved: cannot find dependency uci-firewall for luci-app-firewall
pkg_hash_fetch_best_installation_candidate: Packages for luci-app-firewall found, but incompatible with the architectures configured
satisfy_dependencies_for: Cannot satisfy the following dependencies for luci:
uci-firewall
opkg_install_cmd: Cannot install package luci.
```
how fix it?
thank you

Hurricos commented:
@chunkeey, @mans0n or @adschm: could you take a look at this device support PR?

Please let me know if you'd like me to refresh it to master.

mans0n commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -15,6 +15,17 @@ define Device/aerohive_hiveap-330
 endef
 TARGET_DEVICES += aerohive_hiveap-330
 
+define Device/extreme-networks_ws-ap3825i
```
This should go below `enterasys_ws-ap3710i`.
(BTW it seems Enterasys and Extreme Networks are the same vendor, right?)

Hurricos replied:
Enterasys (and their designs) were purchased by Extreme Networks during the transition; this AP is identical to the Enterasys WS-AP3715i, but with the ath9k 5GHz card replaced with an ath10k one.

Hurricos replied:
Fixed in force-push 557192907b9652b21f1cc8890b807edae22aed40.

mans0n commented on target/linux/mpc85xx/p1020/config-default:
```diff
@@ -8,6 +8,7 @@ CONFIG_EEPROM_LEGACY=y
 CONFIG_GENERIC_CLOCKEVENTS_BROADCAST=y
 CONFIG_GENERIC_TBSYNC=y
 # CONFIG_GPIO_MAX77620 is not set
+CONFIG_GPIO_74X164=y
```
This should go above `CONFIG_GPIO_MAX77620`.

Hurricos replied:
Fixed in force-push 557192907b9652b21f1cc8890b807edae22aed40.

mans0n reviewed (commented):
Sorry, I can't really review on this target, except for mere cosmetic issues:

mans0n commented:
And I'm not a big fan of using pastebin in the commit message as they tend to be short-lived...

Hurricos commented:
> And I'm not a big fan of using pastebin in the commit message as they tend to be short-lived...

I have pulled the pastebin out into a comment on the forum and will replace the "5" link with one to here: https://forum.openwrt.org/t/adding-openwrt-support-for-ws-ap3825i/101168/26

chunkeey commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ws-ap3825i.dts:
```diff
@@ -0,0 +1,246 @@
+	};
+
+	led_spi {
+		/* This is currently non-functioning because the spi-gpio
+		driver refuses to register when presented with this node. */
```
so none of the LEDs are working?

chunkeey commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/ws-ap3825i.dts:
```diff
@@ -0,0 +1,246 @@
+		reset {
+			label = "Reset button";
+			gpios = <&gpio0 8 GPIO_ACTIVE_LOW>;
+			linux,code = <0x198>; /* KEY_RESTART */
```
#include <dt-bindings/input/input.h> and set linux,code = <KEY_RESTART>; ?

(Just a Q: this reset button is on another GPIO. But it does work, right?)

Hurricos replied:
Yes, I confirmed originally and reconfirmed today, this is the GPIO for the reset button and that it does act as such with this node on this GPIO line.

Hurricos replied:
Fixed in force-push 557192907b9652b21f1cc8890b807edae22aed40.

chunkeey commented on target/linux/mpc85xx/image/p1020.mk:
```diff
@@ -15,6 +15,17 @@ define Device/aerohive_hiveap-330
+  DEVICE_MODEL := WS-AP3825i
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
+  BLOCKSIZE := 128k
+  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
```
In the commit message above, you mention

> The flattened device tree must be resized multiple times to deal with
  U-boot's pickiness about making space for the fdt without relocating
  it. Although `help fdt` advertises that `fdt resize` pads to a 4K-aligned
  address, it cannot predict the future size of the fdt after later `fdt`
  operations.

This sounds a lot like the troubles I had with apm821xx, uboot couldn't add/extend nodes.

The solution there was to tell the dtc compiler to pad the dtb with empty space.

apm821xx uses a special setup for this. It overwrites the "dtb" with its own version that passes a new DEVICE_VAR DTB_SIZE which does the rest. (the 32KiB is a bit too generous.)

https://github.com/openwrt/openwrt/blob/master/target/linux/apm821xx/image/Makefile#L31

Maybe this works here as well and you can get rid of the ranges that uboot is supposed to fill?

chunkeey replied:
Any luck with that method? Now that the AP-350 is "out of the way"?

Hurricos replied:
I'm attempting a compile now (this is mostly a comprehension test for me on how image recipe definition lines work; we'll see if I pass).

Hurricos replied:
It works! I need to refine things a bit, but here's what I tried so far:
```diff
diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
index 2b81884032..e1d1bc6948 100644
--- a/target/linux/mpc85xx/image/Makefile
+++ b/target/linux/mpc85xx/image/Makefile
@@ -3,6 +3,12 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/image.mk
 
+DEVICE_VARS += DTB_SIZE
+
+define Build/dtb
+       $(call Image/BuildDTB,$(DTS_DIR)/$(DEVICE_DTS).dts,$@.dtb,,--space $(DTB_SIZE))
+endef
+
 define Device/Default
   PROFILES := Default
   DEVICE_DTS := $(lastword $(subst _, ,$(1)))
diff --git a/target/linux/mpc85xx/image/p1020.mk b/target/linux/mpc85xx/image/p1020.mk
index 93fd315e36..feb327036d 100644
--- a/target/linux/mpc85xx/image/p1020.mk
+++ b/target/linux/mpc85xx/image/p1020.mk
@@ -63,7 +63,8 @@ define Device/extreme-networks_ws-ap3825i
   DEVICE_MODEL := WS-AP3825i
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
   BLOCKSIZE := 128k
-  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
+  DTB_SIZE := 20480
+  KERNEL = dtb | kernel-bin | lzma | dtb | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
   IMAGES := sysupgrade.bin
   IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
 endef
```

Hurricos replied:
A combination of

```txt
interrupts off;
bootm start 0x2000000;
bootm loados;
fdt resize;
fdt boardsetup;
fdt chosen;
bootm prep;
bootm go;
```

... and
```diff
diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
index 2b81884032..e1d1bc6948 100644
--- a/target/linux/mpc85xx/image/Makefile
+++ b/target/linux/mpc85xx/image/Makefile
@@ -3,6 +3,12 @@
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/image.mk
 
+DEVICE_VARS += DTB_SIZE
+
+define Build/dtb
+       $(call Image/BuildDTB,$(DTS_DIR)/$(DEVICE_DTS).dts,$@.dtb,,--space $(DTB_SIZE))
+endef
+
 define Device/Default
   PROFILES := Default
   DEVICE_DTS := $(lastword $(subst _, ,$(1)))
diff --git a/target/linux/mpc85xx/image/p1020.mk b/target/linux/mpc85xx/image/p1020.mk
index 93fd315e36..5049847cb8 100644
--- a/target/linux/mpc85xx/image/p1020.mk
+++ b/target/linux/mpc85xx/image/p1020.mk
@@ -63,7 +63,8 @@ define Device/extreme-networks_ws-ap3825i
   DEVICE_MODEL := WS-AP3825i
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca988x-ct
   BLOCKSIZE := 128k
-  KERNEL = kernel-bin | lzma | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
+  DTB_SIZE := 20480
+  KERNEL = kernel-bin | lzma | dtb | fit lzma $(KDIR)/image-$$(DEVICE_DTS).dtb
   IMAGES := sysupgrade.bin
   IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata
 endef
```

... is enough to avoid repeated resize. I will adjust my commit to match.

Hurricos replied:
Fixed, and pushed in force-push 557192907b9652b21f1cc8890b807edae22aed40.

eutampieri commented:
Are you sure that the shift register is SPI? [Here](https://www.ti.com/lit/ds/symlink/sn74lv164a.pdf) doesn’t seem so.

Apologies if I’m completely mistaken

Hurricos commented:
> Are you sure that the shift register is SPI? [Here](https://www.ti.com/lit/ds/symlink/sn74lv164a.pdf) doesn’t seem so.
> 
> Apologies if I’m completely mistaken

That's the first spec sheet I saw too! I'm not sure what it's supposed to be if not SPI -- what would it be instead, just ... UART?

I was so sure after injecting some messages into the code that loads the driver that the problem was simply the driver not loading.

eutampieri commented:
> Edit: OK, it's not explicitly SPI, but it should work _with_ an SPI master. It's sort of overkill to bit-bang SPI and then use the result to run something that doesn't require full SPI, but I'm not particularly good at writing new drivers.

If I’m not mistaken the driver should be up streamed to mainline Linux. If so, I think it could be an interesting (for me, if I have time ;)) project to write a Linux driver for the shift register

eutampieri commented:
> the problem was simply the driver not loading.

Is this debuggable? How could I start?

Hurricos commented:
> > the problem was simply the driver not loading.
> 
> Is this debuggable? How could I start?

It's most likely the debug process for this one would be to build with printk debug messages, as per this part of the commit:

> LEDs do not function under OpenWrt. Each of 8 LEDs is connected to an
output of a TI LV164A shift register, which is wired to GPIO lines and
operates through bit-banged SPI. Unfortunately, I am unable to get the
spi-gpio driver to recognize the led_spi device tree node at all, as
confirmed by patching in printk messages demonstrating
spi-gpio.c::spi_gpio_probe never runs. It is possible to manually
articulate the shift register by exporting the GPIO lines and stepping
their values through the sysfs.

I am fairly sure `spi-gpio.c::spi_gpio_probe` is supposed to be run as the kernel is reading through and setting up devices specified in the device tree; I might be wrong, though. If I am, we can take another shot. The fact that it does not run to create a bit-banged SPI device on the GPIO line connected to the LV164A is at least one reason for which the shift register isn't working.

Hurricos commented:
In force-push 557192907b9652b21f1cc8890b807edae22aed40, I have
- rebased on top of a much more recent master (after the HiveAP 330 fixes)
- fixed up a number of different outstanding formatting issues
- Removed the need for multiple `fdt resize;`s using @chunkeey's builder dtb padding recipe

chunkeey commented:
> Removed the need for multiple fdt resize;s using @chunkeey's builder dtb padding recipe

glad that it did at least removed some of the `fdt resize`. Though, that SPI-LED is still a mystery.

Hurricos commented:
> > Removed the need for multiple fdt resize;s using @chunkeey's builder dtb padding recipe
> 
> glad that it did at least removed some of the `fdt resize`. Though, that SPI-LED is still a mystery.

Some thoughts:
- I might be missing a kernel symbol somewhere. I didn't create the updates to the kernel config by using a kernel `make menuconfig`, so perhaps the dependencies are off.
  - This feels like the best explanation for why the applicable device tree node is never handled by `spi_gpio_probe`.
- If all else fails, I want to empower a future committer to suggest a fix:
  - that's why I leave extra commentary,
  - and why I'm leaving the relevant dt nodes in.
   
I don't think LED functionality should be a blocker, though. It's not really critical, and it's easy to fix / test in place if/once someone finds the fix.

chunkeey commented:
well, let's find out then. Merged. Cheers!
